### PR TITLE
Implement wp-admin redirects for expired eCommerce trials

### PR DESCRIPTION
**Pull request by daledupreez**
### Changes proposed in this Pull Request:

This PR ensures that when users have an expired free eCommerce trial and visit a page in wp-admin, we will correctly redirect them to the expired trial page in Calypso.

Closes:
* #1005

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

### How to test the changes in this Pull Request:

- [ ] Have you followed the [Writing high-quality testing instructions guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)?

* Set up a WoA dev site that has a _valid_ free eCommerce trial plan - you can do this by following the instructions here: peapX7-1D4-p2
* Upload the changes from this PR to your site -- the code is under `wp-content/mu-plugins/wpcomsh/vendor/automattic/wc-calypso-bridge/`
* Navigate to various pages in the UI, especially in wp-admin, and verify that you don't get any errors, and see the pages you are trying to access
* Use Store Admin to modify the expiration date of your free trial plan to be a date in the past (ideally 1-3 days in the past)
* Wait a few moments, and then try to visit various pages in wp-admin for your site e.g. `/wp-admin/`, `/wp-admin/edit.php`, `/wp-admin/admin.php?page=wc-admin`
* Verify that you are redirected to the expired trial page in Calypso for any `/wp-admin/` page you try to visit
* Now upgrade the plan for your site
 - Note that the post-purchase code may redirect you back to the expired trial site - see Automattic/calypso#74785
* Repeat the tests for various wp-admin pages in your site
* Verify that you can access them all again

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: includes/free-trial/class-wc-calypso-bridge-free-trial-expired-plan-redirects.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+		$trial_plan_purchases = array_filter(
+			$site_purchases,
+			function ( $site_purchase ) {
+				return 'ecommerce-trial-bundle-monthly' === ( $site_purchase['product_slug'] ?? null );
+			}
+		);
+
+		// TODO: if we don't have a trial plan purchase, should we still redirect?
+		if ( empty( $trial_plan_purchases ) ) {
```
Comment (xristos3490) [inline includes/free-trial/class-wc-calypso-bridge-free-trial-expired-plan-redirects.php:76]: Given that the class will bail early in the constructor if there are no ecom features, the purchases array shouldn't be empty. If it is, for some reason, then it's probably something wrong with the system altogether, so just `return`ing here seems correct.

File: includes/free-trial/class-wc-calypso-bridge-free-trial-expired-plan-redirects.php [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	}
+
+	/**
+	 * When we are handling an incoming WP Admin UI request
+	 * AND the site has an expired eCommerce trial,
+	 * redirect to the expired trial page in Calypso.
+	 *
+	 * @return void
+	 */
+	public function maybe_redirect_wp_admin_to_expired_plan_page() {
```
[the remark this answers is not in the document]
Comment (daledupreez) [reply includes/free-trial/class-wc-calypso-bridge-free-trial-expired-plan-redirects.php:49]: Thanks for the suggestion! I have now added the checks.

File: includes/free-trial/class-wc-calypso-bridge-free-trial-expired-plan-redirects.php (hunk 1 above)
Comment (daledupreez) [reply includes/free-trial/class-wc-calypso-bridge-free-trial-expired-plan-redirects.php:82]: I've updated the comment to reflect that this is an unexpected case where we won't take action.

File: includes/free-trial/class-wc-calypso-bridge-free-trial-expired-plan-redirects.php [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+		}
+
+		$trial_plan_purchase = array_shift( $trial_plan_purchases );
+
+		$current_timestamp = wpcom_datetime_to_iso8601( 'now' );
+
+		if (
+			$trial_plan_purchase
+			&& isset( $trial_plan_purchase['expiry_time'] )
+			&& $trial_plan_purchase['expiry_time'] < $current_timestamp
```
[the remark this answers is not in the document]
Comment (daledupreez) [reply includes/free-trial/class-wc-calypso-bridge-free-trial-expired-plan-redirects.php]: I've added a guard for this.

**Review by markbiek — APPROVED** [review]
Working as expected!

**Pull request merged**